File: part-3-cache-mock/python/range.py
class Range:
    def __init__(self, offset: int, size: int):
        self.offset = offset
        self.size = size

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.offset == other.offset and self.size == self.size
        else:
            return False

    def __repr__(self):
        return f"{self.offset}:{self.size}"

    def end(self) -> int:
        return self.offset + self.size

    def overlap(range1, range2) -> bool:
        return not (range1.end() <= range2.offset or range2.end() <= range1.offset)

    def exclude(range1, range2):
        assert(Range.overlap(range1, range2))
        offset_left = min(range1.offset, range2.offset)
        offset_right = max(range1.offset, range2.offset)
        left = Range(offset_left, offset_right - offset_left)
        end_left = min(range1.end(), range2.end())
        end_right = max(range1.end(), range2.end())
        right = Range(end_left, end_right - end_left)
        return left, right

    def intersect(range1, range2):
        assert(Range.overlap(range1, range2))
        offset = max(range1.offset, range2.offset)
        end = min(range1.end(), range2.end())
        return Range(offset, end - offset)

    def shift(self, delta_offset: int):
        return Range(self.offset + delta_offset, self.size)

File: part-3-cache-mock/cpp/range.cpp
/*********************************************************/
#include <cassert>
#include <algorithm>
#include "range.hpp"

/*********************************************************/
using namespace cache;

/*********************************************************/
Range::Range(size_t offset, size_t size)
      :offset(offset)
      ,size(size)
{
}

/*********************************************************/
bool Range::operator==(const Range & other) const
{
	return this->offset == other.offset && this->size == other.size;
}

/*********************************************************/
bool Range::operator<(const Range & other) const
{
	return this->offset < other.offset;
}

/*********************************************************/
size_t Range::end(void) const
{
	return this->offset + this->size;
}

/*********************************************************/
bool Range::overlap(const Range & range1, const Range & range2)
{
	return ! (range1.end() <= range2.offset || range2.end() <= range1.offset);
}

/*********************************************************/
std::pair<Range, Range> Range::exclude(const Range & range1, const Range & range2)
{
	assert(Range::overlap(range1, range2));
	size_t offset_left = std::min(range1.offset, range2.offset);
	size_t offset_right = std::max(range1.offset, range2.offset);
	Range left(offset_left, offset_right - offset_left);
	size_t end_left = std::min(range1.end(), range2.end());
	size_t end_right = std::max(range1.end(), range2.end());
	Range right(end_left, end_right - end_left);
	std::pair<Range, Range> res;
	res.first = left;
	res.second = right;
	return res;
}

/*********************************************************/
Range Range::intersect(const Range & range1, const Range & range2)
{
	assert(Range::overlap(range1, range2));
	size_t offset = std::max(range1.offset, range2.offset);
	size_t end = std::min(range1.end(), range2.end());
	return Range(offset, end - offset);
}
